SmbusNCT6793.p

```diff
@@ -0,0 +1,479 @@
+                    {
+                        for(i = cnt; i <= len; i++)
+                        {
+                            io_out_byte(SMBHSTDAT, in[i]);
+                        }
+
+                        len = 0;
+                    }
+
+                    io_out_byte(SMBHSTCMD, NCT6793_WRITE_BLOCK);
+                }
+                else
+                {
+                    return STATUS_NOT_SUPPORTED;
+                }
+            }
+        
+        default:
+            {
+                return STATUS_NOT_SUPPORTED;
+            }
+    }
+
+    io_out_byte(SMBHSTCTL, NCT6793_MANUAL_START);
+
+    while(size == I2C_SMBUS_BLOCK_DATA && len > 0)
+    {
+        if(read_write == I2C_SMBUS_WRITE)
+        {
+            timeout = 0;
+            while((io_in_byte(SMBHSTSTS) & NCT6793_FIFO_EMPTY) == 0)
+            {
+                if(timeout > MAX_RETRIES)
+                {
+                    return STATUS_TIMEOUT;
+                }
+
+                microsleep(250);
+                timeout++;
+            }
+
+            /* Load more bytes into FIFO */
+            if(len >= 4)
+            {
+                for(i = cnt; i <= (cnt + 4); i++)
+                {
+                    io_out_byte(SMBHSTDAT, in[i]);
+                }
+
+                len -= 4;
+                cnt += 4;
+            }
+            else
+            {
+                for(i = cnt; i <= (cnt + len); i++)
+                {
+                    io_out_byte(SMBHSTDAT, in[i]);
+                }
+
+                len = 0;
+            }
+        }
+    }
+
+    /* Wait for manual mode to complete */
+    timeout = 0;
+    while((io_in_byte(SMBHSTSTS) & NCT6793_MANUAL_ACTIVE) != 0)
+    {
+        if(timeout > MAX_RETRIES)
+        {
+            return STATUS_TIMEOUT;
+        }
+
+        microsleep(250);
+        timeout++;
+    }
+
+    if((io_in_byte(SMBHSTERR) & NCT6793_NO_ACK) != 0)
+    {
+        return STATUS_NO_SUCH_DEVICE;
+    }
+
+    if(read_write == I2C_SMBUS_WRITE)
+    {
+        return STATUS_SUCCESS;
+    }
+
+    switch(size)
+    {
+        case I2C_SMBUS_BYTE_DATA:
+            {
+                out[0] = io_in_byte(SMBHSTDAT);
+            }
+
+        case I2C_SMBUS_WORD_DATA:
+            {
+                out[0] = io_in_byte(SMBHSTDAT);
+                out[1] = io_in_byte(SMBHSTDAT);
+            }
+    }
+
+    return STATUS_SUCCESS;
+}
+
+/// Identify the SMBus controller.
+///
+/// @param out [0] = Type of the SMBus controller, [1] = I/O Base address, [2] = PCI vendor/device ID, subsystem vendor/device ID
+/// @param out_size Must be 3
+/// @return An NTSTATUS
+forward NTSTATUS:ioctl_identity(in[], in_size, out[], out_size);
+public NTSTATUS:ioctl_identity(in[], in_size, out[], out_size)
+{
+    if (out_size < 3)
+    {
+        return STATUS_BUFFER_TOO_SMALL;
+    }
+
+    out[0] = driver_name;
+
+    out[1] = nuvoton_nct6793_smba;
+
+    // Not a PCI device, set PCI information to zero
+    out[2] = 0;
```

Comment:
Vendor id 0000 is assigned to **Gammagraphx, Inc**. I think you're looking for VID:DID of FFFF:FFFF which is the no response value from bus master.

Reply:
Good catch, I've updated it.